cluster-autoscaler/core/scaledown/actuation/actuator.go

```diff
@@ -197,17 +217,90 @@ func (a *Actuator) deleteAsyncDrain(drain []*apiv1.Node) (scaledDownNodes []*sta
 			continue
 		}
+
 		a.nodeDeletionTracker.StartDeletionWithDrain(nodeGroup.Id(), drainNode.Name)
-		go a.scheduleDeletion(drainNode, nodeGroup.Id(), true)
+		groupIds = append(groupIds, nodeGroup.Id())
+		validNodes = append(validNodes, drainNode)
 	}
+
+	go a.deleteNodesAsync(validNodes, groupIds, true)
+
 	return scaledDownNodes
 }
 
+func (a *Actuator) deleteNodesAsync(nodes []*apiv1.Node, groupIds []string, drain bool) {
+	var pdbs []*policyv1.PodDisruptionBudget
+
+	if *nodeDeleteDelayAfterTaint > time.Duration(0) {
+		klog.V(0).Infof("Scale-down: waiting %v before trying to delete nodes", *nodeDeleteDelayAfterTaint)
+		time.Sleep(*nodeDeleteDelayAfterTaint)
+	}
+
+	clusterSnapshot, err := a.createSnapshot(nodes)
+
+	if err != nil {
+		klog.Errorf("Scale-down: couldn't create delete snapshot, err: %v", err)
+		nodeDeleteResult := status.NodeDeleteResult{ResultType: status.NodeDeleteErrorInternal, Err: errors.NewAutoscalerError(errors.InternalError, "createSnapshot returned error %v", err)}
+		for i, node := range nodes {
+			CleanUpAndRecordFailedScaleDownEvent(a.ctx, node, groupIds[i], drain, a.nodeDeletionTracker, "failed to create delete snapshot", nodeDeleteResult)
+		}
+		return
+	}
+
+	if drain {
+		pdbs, err = a.ctx.PodDisruptionBudgetLister().List()
+		if err != nil {
+			klog.Errorf("Scale-down: couldn't fetch pod disruption budgets, err: %v", err)
+			nodeDeleteResult := status.NodeDeleteResult{ResultType: status.NodeDeleteErrorInternal, Err: errors.NewAutoscalerError(errors.InternalError, "podDisruptionBudgetLister.List returned error %v", err)}
+			for i, node := range nodes {
+				CleanUpAndRecordFailedScaleDownEvent(a.ctx, node, groupIds[i], drain, a.nodeDeletionTracker, "failed to fetch pod disruption budgets", nodeDeleteResult)
+			}
+			return
+		}
+	}
+
+	for i, node := range nodes {
+		nodeInfo, err := clusterSnapshot.NodeInfos().Get(node.Name)
+		if err != nil {
+			klog.Errorf("Scale-down: can't retrieve node %q from snapshot, err: %v", node.Name, err)
+			nodeDeleteResult := status.NodeDeleteResult{ResultType: status.NodeDeleteErrorInternal, Err: errors.NewAutoscalerError(errors.InternalError, "nodeInfos.Get for %q returned error: %v", node.Name, err)}
+			CleanUpAndRecordFailedScaleDownEvent(a.ctx, node, groupIds[i], drain, a.nodeDeletionTracker, "failed to get node info", nodeDeleteResult)
+			continue
+		}
+
+		if drain {
+			if _, _, _, err := simulator.GetPodsToMove(nodeInfo, a.ctx.ListerRegistry, pdbs, time.Now()); err != nil {
+				nodeDeleteResult := status.NodeDeleteResult{ResultType: status.NodeDeleteErrorInternal, Err: errors.NewAutoscalerError(errors.InternalError, "GetPodsToMove for %q returned error: %v", node.Name, err)}
+				CleanUpAndRecordFailedScaleDownEvent(a.ctx, node, groupIds[i], drain, a.nodeDeletionTracker, "failed to get pods to move on node", nodeDeleteResult)
+				continue
+			}
+		} else {
+			podsToRemove, _, _, err := simulator.FastGetPodsToMove(nodeInfo, true, true, nil, time.Now())
```

Comment:
`FastGetPodsToMove` and `DetailedGetPodsToMove` aren't really different: they used to be different long time ago before informers were used to keep cached lists of kubernetes objects up to date (the detailed version used to make API calls back then). I'd just use a single call for both empty and drained here to simplify the code a little.

Reply:
Fixed as well, changed to call only DetailedGetPodsToMove.